### PR TITLE
Fault Tolerant Manual: Add _SupportsStateDict to validate a class is stateful

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Find the complete PoC here: https://github.com/PyTorchLightning/pytorch-lightning/pull/10568

Parts of #10431

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @justusschock @awaelchli @ninginthecloud

### CODE REVIEW

File: pytorch_lightning/utilities/auto_restart.py [hunk 1]
```diff
@@ -571,3 +572,15 @@ def reload_dataloader_state_dict(dataloader: DataLoader, state_dict: Dict[str, A
 
     else:
         raise MisconfigurationException("This shouldn't happen. Please, open an issue on PyTorch Lightning Github.")
+
+
+def is_obj_stateful(obj: Any) -> bool:
```
Comment (ananthsub) [inline pytorch_lightning/utilities/auto_restart.py]: Python 3.8 supports Protocols which you might find helpful here: https://www.python.org/dev/peps/pep-0544/

Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py]: I checked it out and it seemed an overkill.

Comment (carmocca) [reply pytorch_lightning/utilities/auto_restart.py]: I think that's a great suggestion which is also Pythonic and is designed exactly for the check you want to implement.

We are not limited by 3.8 because it's available in `typing_extensions` which we already install

Comment (carmocca) [reply pytorch_lightning/utilities/auto_restart.py]: Pushed d04596d

File: CHANGELOG.md [hunk 2]
```diff
@@ -13,6 +13,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 
 - Fault Tolerant Manual
+    * Add `_is_obj_stateful` utility to detect if user data loading components are stateful ([#10646](https://github.com/PyTorchLightning/pytorch-lightning/issues/10646))
```
Comment (carmocca) [inline CHANGELOG.md:16]:
```suggestion
    * Add `_SupportsStateDict` protocool to detect if components are stateful ([#10646](https://github.com/PyTorchLightning/pytorch-lightning/issues/10646))
```

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/utilities/auto_restart.py [hunk 3]
```diff
@@ -570,3 +572,15 @@ def reload_dataloader_state_dict(dataloader: DataLoader, state_dict: Dict[str, A
 
     else:
         raise MisconfigurationException("This shouldn't happen. Please, open an issue on PyTorch Lightning Github.")
+
+
+def _is_obj_stateful(obj: Any) -> bool:
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py]: Removed !

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/utilities/auto_restart.py [hunk 4]
```diff
@@ -570,3 +571,12 @@ def reload_dataloader_state_dict(dataloader: DataLoader, state_dict: Dict[str, A
 
     else:
         raise MisconfigurationException("This shouldn't happen. Please, open an issue on PyTorch Lightning Github.")
+
+
+@runtime_checkable
+class _SupportsStateDict(Protocol):
+    def state_dict(self) -> Dict[str, Any]:
+        ...
```
Comment (Borda) [inline pytorch_lightning/utilities/auto_restart.py:579]: rather add a docstring and explain what it shall do...

Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:582]: This is a protocol class, so better to use ... and it is for private usage.

File: pytorch_lightning/utilities/auto_restart.py [hunk 5]
```diff
@@ -570,3 +571,15 @@ def reload_dataloader_state_dict(dataloader: DataLoader, state_dict: Dict[str, A
 
     else:
         raise MisconfigurationException("This shouldn't happen. Please, open an issue on PyTorch Lightning Github.")
+
+
+@runtime_checkable
+class _SupportsStateDict(Protocol):
+
```
Comment (awaelchli) [inline pytorch_lightning/utilities/auto_restart.py:578]:
```suggestion
```

File: tests/utilities/test_auto_restart.py [hunk 6]
```diff
@@ -1195,6 +1196,29 @@ def test_auto_restart_under_signal(on_last_batch, val_check_interval, failure_on
         assert "dataloader_state_dict" in state_dict
 
 
+def test_supports_state_dict_protocol():
```
Comment (awaelchli) [inline tests/utilities/test_auto_restart.py:1199]: If I understand correctly, it is also going to fail for incompatible signature, correct?

**Review by awaelchli — APPROVED** [review]

File: tests/utilities/test_auto_restart.py (hunk 6 above)
Comment (tchaton) [reply tests/utilities/test_auto_restart.py:1199]: isinstance would return False with incompatible signature.
